### ISSUE TITLE
React-like context

### ISSUE THREAD

**Issue by Aetet** [angular/angular#4698]
Hello.
For data-handling frameworks will be extremely usefull if angular2 will got some context, which will be accessible from any widget in a tree. 

For example [redux](https://www.npmjs.com/package/ng2-redux). With current implementation we must pass state in all widgets like this https://github.com/cascada-io/island-angular2/blob/master/components/day/tmpl.html#L3

It is not very common to pass `m` and `actions` everywhere from top to bottom, so in React this problem solved with internal context, so how this problem will be solved in angular2?

**Comment by gdi2290** [comment]
di

**Comment by tbosch** [comment]
@gdi2290 could you add an example for @Aetet :-)

**Comment by Aetet** [comment]
oh, that would be great.

**Comment by bradlygreen** [comment]
@mhevery can you take a look at this request?  seems like a good boilerplate savings if there's a reasonable method for making it work.

**Comment by mhevery** [comment]
@Aetet from what I can tell, it seems to me that you would like an object called 'context' to be available to all components. Can't you just configure it in the DI and then pass it in?

```
class Context {}

@Component({
  selector: 'foo',
  template: '...'
})
class Foo {
  constructor(context: Context)
}

@Component({
  provider(Context),
  template: '<foo></foo>',
  directives: [Foo]
})
class MyApp{
}

bootstrap(MyApp);
```

**Comment by Aetet** [comment]
Ok, I see. That's exactly what I need. Thanks. And one more question where I can read about them more in a doc? Currently doc preview API responds with 404.

**Comment by robwormald** [comment]
@Aetet re: redux - i did a demo of this a few weeks back - http://plnkr.co/edit/wj4lRtAwNMry8ToNzLrm

tldr: redux store wrapped in an observable stream, pushing updates to the components. you might find it helpful.

**Comment by mhevery** [comment]
working as intended

**Comment by mmavko** [comment]
Hello,

how about the actual problem React Context API is meat to solve?

> In a typical React application, data is passed top-down (parent to child) via props, but this can be cumbersome for certain types of props (e.g. locale preference, UI theme) that are required by many components within an application. Context provides a way to share values like these between components without having to explicitly pass a prop through every level of the tree.

> (https://reactjs.org/docs/context.html)

I would like to write something like this in my template:

```
<ngContext value="dark"><Toolbar></Toolbar></ngContext>
<ngContext value="light"><Toolbar></Toolbar></ngContext>
```

where `Toolbar` is a component which itself renders for example `ToolbarButton` which depends on context value.

What's a canonical solution for this problem in Angular?

**Comment by mmavko** [comment]
@mlc-mlapis it doesn't really matter. The point is that a value is passed some way down through the component hierarchy. And this value is defined in the template. And a value can be different for different regions in the same template.

**Comment by mlc-mlapis** [comment]
@mmavko ... so how such a context is different from the context of a component which is used in Angular today?

**Comment by mmavko** [comment]
@mlc-mlapis what do you mean by "context of a component"?

**Comment by mlc-mlapis** [comment]
@mmavko ... yep. The example is a bit different because it a structural directive is used but it doesn't matter ... there is still an instance of some class and all of its properties and methods define that context.

**Comment by mmavko** [comment]
This doesn't really address the problem. How do you define a context for `ToolbarButton` in your `AppComponent`? If you use structural directive for a component you can somehow influence behavior of its _direct_ children (since a structural directive creates a view and adds child components "manually"). What about grandchildren, which are rendered by children?

**Comment by mlc-mlapis** [comment]
@mmavko ... simply `<Toolbar *appContext="value"></Toolbar>` leads to:
```
<ng-template [appContext]="value">
   <Toolbar></Toolbar>
</ng-template>
```

**Comment by gdi2290** [comment]
@mmavko using di is a much better method of doing the same thing as React's context api since you can create scoped context rather than working in a global scope as react's context.

Toolbar could require a parent string token for `"theme"` (we use a string because we don't want to share the same reference unless that's not a problem) then the parent component can create the service with the string token as the key

**Comment by mmavko** [comment]
Thanks you for you comments, guys. May I ask you to provide an example of this?

Here's what I'm doing with React:
https://next.plnkr.co/edit/ESWRJapMh8k9wCe9

And that's a boilerplate for similar thing with Angular:
https://next.plnkr.co/edit/OGnIpgnIsVC9wZXb

If you can fork this last fiddle and show how I can do a similar thing, I would appreciate it. Thank you!
